### PR TITLE
Combine inverse map and log density calculation

### PR DESCRIPTION
**Pull request by jmarshrossney**
Combined the inverse map from base to target distribution with the calculation of the model density.
This avoids going explicitly applying the forward map to calculate the density, as was being done before.

**Justification:**
- Speed, primarily. For the example `phi4_train.yml` runcard, it takes 1.5x longer to train for the same number of epochs when the density is calculated separately, compared to when it is merged into the inverse map. This is true for `RealNVP` for which the map and inverse are very cheap to compute, but the speed gains will be bigger for a more complicated map.
- I think this slightly simplifies the task of combining different modules. For example, we want to sandwich `RealNVP` between a projection map and its inverse, which could be easily implemented in another `torch.nn.Module` which includes a single call to `RealNVP.forward()`, rather than needing to call `inverse_map()` and `forward()` separately.

**Drawbacks:**
- Sampling is now slightly messier, with the current log density needing to be passed between batches in addition to the current state.
- Slight reduction in transparency.

### CODE REVIEW

**Comment by jmarshrossney** [comment]
Also fixed a typo in one of the parsers, which I missed last time, and finally changed the Metropolis algorithm to use `random.random()` to generate random numbers, which is faster than `numpy.uniform()` plus indexing.

**Comment by wilsonmr** [comment]
Seems sensible, I did the double pass when I was still making sure the gradients were being tracked properly and it was all explicit but you're right it's very inefficient.

File: anvil/sample.py [hunk 1]
```diff
@@ -103,9 +104,9 @@ def thermalised_state(loaded_model, action, thermalisation) -> torch.Tensor:
     if thermalisation is None:
         return thermalisation
 
-    states, _ = sample_batch(loaded_model, action, thermalisation)
+    states, current_log_density, _ = sample_batch(loaded_model, action, thermalisation)
     log.info(f"Thermalisation: discarded {thermalisation} configurations.")
-    return states[-1]
+    return states[-1], current_log_density
```
Comment (wilsonmr) [inline anvil/sample.py]: ok I'm still thinking the best way to do this so don't change yet but I think that we should just make sample_batch return the full stack of both states *and* log density and then this function should index the last item for both because then it becomes canon that whenever we have a phi or a stack of phis, we also have a corresponding log_density

Comment (wilsonmr) [reply anvil/sample.py]: i think the symmetry would make it easier to follow

**Comment by wilsonmr** [comment]
hmm maybe this is stupid so bear with:

what if every forward-like pass returned a similar tuple, pretty sure the calls to `log_det_jacobian` are also made redundant by this

Also this could extend to generators and additional layer-like objects so I'm really looking a `tan`/`tanh` outputs here

that way whenever these things are executed sequentially we can always calc the log_density by summing the log_density contribution from all the things

pseudocode:

```python
z, simple_prob = generator()
phi, model_log_jac = model(z)
squashed_phi, squash_log_jac = tanh_layer(phi)
p_tilde = simple_prob + model_log_jac + squash_log_jac
...
```

is this dumb, what do you think?

**Comment by jmarshrossney** [comment]
Actually I think that's a great idea.

It would also make it easier to stack multiple `nn.Module`s without needing to pass a parameter saying "don't include the log of the base distribution" as I've been doing up to now.

File: anvil/models.py [hunk 2]
```diff
@@ -240,7 +234,17 @@ def log_det_jacobian(self, phi_input):
...
+
+    def forward(self, z_input: torch.Tensor) -> tuple:
+        """Given an stack of states z_input, returns a tuple 
+        (phi_out, log_det_jacob) containing the (inversely) transformed
+        states and the log of the jacobian determinant of the forward
+        transformation.
+        """
+        phi_out = self.inverse_coupling_layer(z_input)
+        log_det_jacob = self.log_det_jacobian(phi_out)
+        return phi_out, log_det_jacob
```
Comment (jmarshrossney) [inline anvil/models.py]: @wilsonmr is this roughly what you had in mind? (but with better docstring)

**Comment by jmarshrossney** [comment]
- The base distribution `generator` objects are no longer attributes of the model (which is great!)
- Basically anything that returns a tensor of states also returns the associated log quantity, which can be the log probability density or the log Jacobian determinant from the transformation.
- Renamed inputs for the loss function in anticipation of the target density being not just the action, but including a term from a non-trivial parameterisation. Eventually I would aim to replace most calls to action by calls to `log_target_density` or something similar.

File: anvil/models.py (hunk 2 above)
Comment (wilsonmr) [reply anvil/models.py]: the output is good yeah. Now please check this because if we get this wrong we break the whole thing lol

since phi = [phi_a, phi_b] -> inverse coupling -> [phi_a, z_b] (in other words phi_a is always left untouched by transformation)

then I think that if we had

```python
    def forward(self, z_input):
        z_a = z_input[:, self._a_ind]
        z_b = z_input[:, self._b_ind]
        s_out = self._s_forward(z_a)
        t_out = self._t_forward(z_a)
        phi_b = (z_b - t_out) * torch.exp(-s_out)
        return self.join_func([z_a, phi_b], dim=1), s_out.sum(dim=1, keepdim=True) # EDIT keepdim not keepdims
```

then we would get the same thing, with one less pass through the net.

I made a dummy_forward which was a copy paste of the above and it seems fine to me:

```python
>>> l = AffineLayer(10, (10, 10), (10, 10), 0)
>>> 
>>> import torch
>>> test_tensor = torch.rand(1, 10)
>>> 
>>> l(test_tensor)
(tensor([[0.9308, 0.7645, 0.3430, 0.7738, 0.1795, 1.0089, 0.7677, 0.5743, 0.4620,
         0.2349]], grad_fn=<CatBackward>), tensor([[0.1624]], grad_fn=<SumBackward1>))
>>> l.dummy_forward(test_tensor)
(tensor([[0.9308, 0.7645, 0.3430, 0.7738, 0.1795, 1.0089, 0.7677, 0.5743, 0.4620,
         0.2349]], grad_fn=<CatBackward>), tensor([[0.1624]], grad_fn=<SumBackward1>))
```

and we see it's faster

```python
In [1]: from anvil.models import AffineLayer                                                                                                                                                                

In [2]: import torch                                                                                                                                                                                        

In [3]: tens = torch.rand(1, 10)                                                                                                                                                                            

In [4]: l = AffineLayer(10, (10, 10), (10, 10), 0)                                                                                                                                                          

In [5]: def test_dummy(s, lay): 
   ...:     with torch.no_grad(): 
   ...:         return lay.dummy_forward(s) 
   ...:                                                                                                                                                                                                     

In [6]: def test_forward(s, lay): 
   ...:     with torch.no_grad(): 
   ...:         return lay(s) 
   ...:                                                                                                                                                                                                     

In [7]: %timeit test_dummy(tens, l)                                                                                                                                                                         
135 µs ± 359 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [8]: %timeit test_forward(tens, l)                                                                                                                                                                       
192 µs ± 848 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

and remember that speed up scales with size of the network!

Comment (jmarshrossney) [reply anvil/models.py]: ah great - more gains. I guess we're rediscovering the USP of RealNVP which is that it's dirt cheap.

I will update and run some checks.

Comment (jmarshrossney) [reply anvil/models.py]: oops @ keepdims

**Comment by jmarshrossney** [comment]
Shall we keep `AffineLayer.coupling_layer` despite the fact that it is currently unused? At some point we may want to actually test that we can recover the simple base distribution from our model output.

**Comment by wilsonmr** [comment]
Yeah exactly, doesn't hurt to keep it

File: anvil/models.py (hunk 2 above)
Comment (wilsonmr) [reply anvil/models.py]: they purposely made tensors behave like numpy arrays, not sure why they changed all of the nomenclature!

**Comment by wilsonmr** [comment]
Is this ready for me to give a final look over?

**Comment by jmarshrossney** [comment]
It agrees with the "old" version as far as I can tell and passes the free theory tests from the adjacent PR, so I think it's good to go!

File: anvil/models.py (hunk 2 above)
Comment (wilsonmr) [reply anvil/models.py]: @marshrossney did you change anything here? I don't see it

**Review by wilsonmr — COMMENTED** [review]
see comment on forward of the affinelayer otherwise looks good

**Comment by jmarshrossney** [comment]
I didn't actually push, my mistake. Sorry!

**Review by wilsonmr — APPROVED** [review]
